test/e2e/upgrade_management_cluster.go

```diff
@@ -0,0 +1,137 @@
+package e2e
+
+import (
+	"fmt"
+	"html/template"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/config"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+)
+
+func upgradeManagementCluster() {
+	var (
+		specName            = "clusterctl-upgrade"
+		upgradeClusterProxy framework.ClusterProxy
+	)
+	upgradeClusterProxy = bootstrapClusterProxy.GetWorkloadCluster(ctx, cluster.Namespace, cluster.Name)
+
+	clusterctlBinaryURLTemplate := e2eConfig.GetVariable("INIT_WITH_BINARY")
+	clusterctlBinaryURLReplacer := strings.NewReplacer("{OS}", runtime.GOOS, "{ARCH}", runtime.GOARCH)
+	clusterctlBinaryURL := clusterctlBinaryURLReplacer.Replace(clusterctlBinaryURLTemplate)
+
+	Logf("Downloading clusterctl binary from %s", clusterctlBinaryURL)
+	clusterctlBinaryPath := downloadToTmpFile(clusterctlBinaryURL)
+	defer os.Remove(clusterctlBinaryPath) // clean up
+
+	err := os.Chmod(clusterctlBinaryPath, 0744) //nolint:gosec
+	Expect(err).ToNot(HaveOccurred(), "failed to chmod temporary file")
+
+	By("Initializing the workload cluster with older versions of providers")
+
+	contract := "v1alpha4"
```

Comment:
This should probably come from the config file or environment. Maybe use `UPGRADE_FROM_CAPI_VERSION`?